src/Optimization/hiopAlgPrimalDecomp.cpp

```diff
@@ -0,0 +1,1211 @@
+  int ret = MPI_Comm_size(MPI_COMM_WORLD, &comm_size_); assert(ret==MPI_SUCCESS);
+  if(my_rank_==0) { 
+    my_rank_type_ = 0;
+  } else {
+    my_rank_type_ = 1;
+  }
+  request_ = new MPI_Request[4];   
+#endif
+  x_ = new double[n_];
+}
+
+
+hiopAlgPrimalDecomposition::~hiopAlgPrimalDecomposition()
+{
+  delete [] x_;
+}
+
+
+double hiopAlgPrimalDecomposition::getObjective() const
+{
+  return master_prob_->get_objective();
+}
+
+void hiopAlgPrimalDecomposition::getSolution(double* x) const
+{
+  master_prob_->get_solution(x);
+}
+  
+void hiopAlgPrimalDecomposition::getDualSolutions(double* zl, double* zu, double* lambda)
+{
+  assert(false && "not implemented");
+}
+
+inline hiopSolveStatus hiopAlgPrimalDecomposition::getSolveStatus() const
+{
+  return solver_status_;
+}
+
+int hiopAlgPrimalDecomposition::getNumIterations() const
+{
+  assert(false && "not yet implemented");
+  return 9;
+}
+  
+bool hiopAlgPrimalDecomposition::stopping_criteria(const int it, const double convg)
+{
+  //gradient based stopping criteria
+  if(convg<tol_){printf("reaching error tolerance, successfully found solution\n"); return true;}
+  //stopping criteria based on the change in objective function
+  if(it == max_iter-1){printf("reached maximum iterations, optimization stops.\n"); return true;}
+
+  if(accp_count == 10){printf("reached acceptable tolerance of %18.12e for 10 iterations, " 
+		              "optimization stops.\n", accp_tol_); return true;}
+  return false;
+}
+  
+double hiopAlgPrimalDecomposition::
+step_size_inf(const int nc, const double* x, const double* x0)
+{
+  double step = -1e20;
+  for(int i=0; i<nc; i++) {
+    step = std::max(step,fabs(x[i]-x0[i]));
+  }
+  return step;
+}
+
+
+void hiopAlgPrimalDecomposition::set_max_iteration(const int max_it)  
+{
+  max_iter = max_it;
+}
+
+void hiopAlgPrimalDecomposition::set_verbosity(const int i)
+{
+  assert(i<=3 && i>=0);
+  ver_ = i;
+}
+
+void hiopAlgPrimalDecomposition::set_tolerance(const double tol)
+{
+  tol_ = tol;
+}
+
+void hiopAlgPrimalDecomposition::set_acceptable_tolerance(const double tol)
+{
+  accp_tol_ = tol;
+}
+
+
+void hiopAlgPrimalDecomposition::set_initial_alpha_ratio(const double alpha)
+{
+  assert(alpha>=0&&alpha<10.);
+  alpha_ratio_ = alpha;
+}
+
+/* MPI engine for parallel solver
+ */
+
+#ifdef HIOP_USE_MPI
+  hiopSolveStatus hiopAlgPrimalDecomposition::run()
+  {
+
+    if(comm_size_==1) {
+      return run_single();//call the serial solver
+    }
+    if(my_rank_==0) {
+      printf("total number of recourse problems  %d\n", S_);
+      printf("total ranks %d\n",comm_size_);
+
+    }
+    //initial point for now set to all zero
+    for(int i=0; i<n_; i++) {
+      x_[i] = 0.;
+    }
+      
+    bool bret;
+    int rank_master=0; //master rank is also the solver rank
+    //Define the values and gradients as needed in the master rank
+    double rval = 0.;
+    double grad_r[nc_];
+    for(int i=0; i<nc_; i++) {
+      grad_r[i] = 0.;
+    }
```

Comment:
Even though the assumption is the master will run on host, the algorithm should be implemented in a generic fashion, not accessing raw data directly.

Reply:
Yes. This now uses hiopVector and calls setToZero().